aws/resource_aws_redshift_snapshot_copy_grant.go

```diff
@@ -0,0 +1,266 @@
+by the describe grants call when expected.
+*/
+
+// NB: This function only retries the grant not being returned and some edge cases, while AWS Errors
+// are handled by the findAwsRedshiftSnapshotCopyGrant function
+func findAwsRedshiftSnapshotCopyGrantWithRetry(conn *redshift.Redshift, grantName string) (*redshift.SnapshotCopyGrant, error) {
+	var grant *redshift.SnapshotCopyGrant
+	err := resource.Retry(3*time.Minute, func() *resource.RetryError {
+		var err error
+		grant, err = findAwsRedshiftSnapshotCopyGrant(conn, grantName, nil)
+
+		if err != nil {
+			if serr, ok := err.(AwsRedshiftSnapshotCopyGrantMissingError); ok {
+				// Force a retry if the grant should exist
+				return resource.RetryableError(serr)
+			}
+
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	return grant, err
+}
+
+// Used by the tests as well
+func waitForAwsRedshiftSnapshotCopyGrantToBeDeleted(conn *redshift.Redshift, grantName string) error {
+	err := resource.Retry(3*time.Minute, func() *resource.RetryError {
+		grant, err := findAwsRedshiftSnapshotCopyGrant(conn, grantName, nil)
+		if err != nil {
+			if isAWSErr(err, redshift.ErrCodeSnapshotCopyGrantNotFoundFault, "") {
+				return nil
+			}
+		}
+
+		if grant != nil {
+			// Force a retry if the grant still exists
+			return resource.RetryableError(
+				fmt.Errorf("[DEBUG] Grant still exists while expected to be deleted: %s", *grant.SnapshotCopyGrantName))
+		}
+
+		return resource.NonRetryableError(err)
+	})
+
+	return err
+}
+
+// The DescribeSnapshotCopyGrants API defaults to listing only 100 grants
+// Use a marker to iterate over all grants in "pages"
+// NB: This function only retries on AWS Errors
+func findAwsRedshiftSnapshotCopyGrant(conn *redshift.Redshift, grantName string, marker *string) (*redshift.SnapshotCopyGrant, error) {
+
+	input := redshift.DescribeSnapshotCopyGrantsInput{
+		MaxRecords: aws.Int64(int64(100)),
+	}
+
+	// marker and grant name are mutually exclusive
+	if marker != nil {
+		input.Marker = marker
+	} else {
+		input.SnapshotCopyGrantName = aws.String(grantName)
+	}
+
+	var out *redshift.DescribeSnapshotCopyGrantsOutput
+	var err error
+	var grant *redshift.SnapshotCopyGrant
+
+	err = resource.Retry(3*time.Minute, func() *resource.RetryError {
+		out, err = conn.DescribeSnapshotCopyGrants(&input)
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	grant = getAwsRedshiftSnapshotCopyGrant(out.SnapshotCopyGrants, grantName)
+	if grant != nil {
+		return grant, nil
+	} else if out.Marker != nil {
+		log.Printf("[DEBUG] Snapshot copy grant not found but marker returned, getting next page via marker: %s", aws.StringValue(out.Marker))
+		return findAwsRedshiftSnapshotCopyGrant(conn, grantName, out.Marker)
+	}
+
+	return nil, NewAwsRedshiftSnapshotCopyGrantMissingError(fmt.Sprintf("[DEBUG] Grant %s not found", grantName))
+}
+
+// Custom error, so we don't have to rely on
```

Comment:
We can remove this in preference of using the existing upstream [`resource.NotFoundError`](https://godoc.org/github.com/hashicorp/terraform/helper/resource#NotFoundError)